libkbfs/bserver_remote.go

```diff
@@ -405,25 +380,39 @@ func (b *BlockServerRemote) batchDowngradeReferences(ctx context.Context,
 			}
 			nonces[BlockRefNonce(ref.Ref.Nonce)] = ref.LiveCount
 		}
-
-		//figure out the not-yet-deleted references
-		oldNotDoneLen := len(notDone)
+		// update the list of references to downgrade
 		notDone = b.getNotDone(contexts, doneRefs)
-		if !madeProgress {
-			madeProgress = len(notDone) < oldNotDoneLen
-		}
-		if !madeProgress {
-			return doneRefs,
-				errors.New("Made no progress after a batch downgrade call")
-		}
 
 		//if context is cancelled, return immediately
 		select {
 		case <-ctx.Done():
-			return doneRefs, ctx.Err()
+			finalError = ctx.Err()
+			return nil
 		default:
 		}
 
+		// check whether to backoff and retry
+		if err != nil {
+			// if error is of type throttle, retry
+			if _, ok := err.(BServerErrorThrottle); ok {
+				return err
+			}
+			// non-throttle error, do not retry here
+			finalError = err
+		}
+		return nil
```

Comment:
Are we guaranteed that the server will only return a `nil` when the completed list matches the `Refs` list exactly? If not, there could be a bug in returning early before everything is complete.  If so, maybe that's worth double checking here that `len(notDone) == 0` when `err == nil`, just to be safe?